lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1353 @@
+            input_shape = self.input_shapes[0]
+            # Explicitly unroll the recurrence instead of using scan
+            hid_out = unroll_scan(
+                fn=step_fun,
+                sequences=sequences,
+                outputs_info=[hid_init],
+                go_backwards=self.backwards,
+                non_sequences=non_seqs,
+                n_steps=input_shape[1])[0]
+        else:
+            # Scan op iterates over first dimension of input and repeatedly
+            # applies the step function
+            hid_out = theano.scan(
+                fn=step_fun,
+                sequences=sequences,
+                go_backwards=self.backwards,
+                outputs_info=[hid_init],
+                non_sequences=non_seqs,
+                truncate_gradient=self.gradient_steps,
+                strict=True)[0]
+
+        # dimshuffle back to (n_batch, n_time_steps, n_features))
+        hid_out = hid_out.dimshuffle(1, 0, 2)
+
+        # if scan is backward reverse the output
+        if self.backwards:
+            hid_out = hid_out[:, ::-1, :]
+
+        self.hid_out = hid_out
+        return hid_out
+
+
+class RecurrentLayer(CustomRecurrentLayer):
+    """
+    lasagne.layers.recurrent.RecurrentLayer(incoming, num_units,
+    W_in_to_hid=lasagne.init.Uniform(), W_hid_to_hid=lasagne.init.Uniform(),
+    b=lasagne.init.Constant(0.), nonlinearity=lasagne.nonlinearities.rectify,
+    hid_init=lasagne.init.Constant(0.), backwards=False, learn_init=False,
+    gradient_steps=-1, grad_clipping=False, unroll_scan=False,
+    precompute_input=True, mask_input=None, **kwargs)
+
+    Dense recurrent neural network (RNN) layer
+
+    A "vanilla" RNN layer, which has dense input-to-hidden and
+    hidden-to-hidden connections.  The output is computed as
+
+    .. math ::
+        h_t = \sigma(x_t W_x + h_{t-1} W_h + b)
+
+    Parameters
+    ----------
+    incoming : a :class:`lasagne.layers.Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    num_units : int
+        Number of hidden units in the layer.
+    W_in_to_hid : Theano shared variable, numpy array or callable
+        Initializer for input-to-hidden weight matrix (:math:`W_x`).
+    W_hid_to_hid : Theano shared variable, numpy array or callable
+        Initializer for hidden-to-hidden weight matrix (:math:`W_h`).
+    b : Theano shared variable, numpy array, callable or None
+        Initializer for bias vector (:math:`b`). If None is provided there will
+        be no bias.
+    nonlinearity : callable or None
+        Nonlinearity to apply when computing new state (:math:`\sigma`). If
+        None is provided, no nonlinearity will be applied.
+    hid_init : callable, np.ndarray, theano.shared or TensorVariable
+        Passing in a TensorVariable allows the user to specify
+        the value of `hid_init` (:math:`h_0`). In this mode `learn_init` is
+        ignored.
```

Comment:
It's a bit confusing that only the `TensorVariable` case is explained. I'd explain what the parameter is for first. Something like this:
```
    hid_init: Theano shared variable, numpy array, callable or TensorVariable
        Initializer for initial hidden state (:math:`h_0`). If set to a TensorVariable
        (Theano expression), it will not be learned regardless of `learn_init`.
```